web-components/src/components/table-advanced/TableAdvanced.ts

```diff
@@ -0,0 +1,994 @@
+import { templateContent } from "lit-html/directives/template-content";
+import { nothing, TemplateResult } from "lit-html";
+import Papa from "papaparse";
+import { customElementWithCheck } from "@/mixins/CustomElementCheck";
+import { Filter } from "./src/filter";
+import { debounce, Evt, evt, TemplateCallback, templateCallback, TemplateInfo } from "./src/decorators";
+
+const IMG = document.createElement("img");
+IMG.src = "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7";
+
+export namespace TableAdvanced {
+  /**
+   * @element md-table-advanced
+   * @fires md-table-advanced-change
+   */
+  @customElementWithCheck("md-table-advanced")
+  export class ELEMENT extends LitElement {
+    @property({ attribute: false }) config!: Config;
+    @property({ attribute: false }) data!: Data;
+
+    @evt() "md-table-advanced-change"!: Evt<ChangeEvent>;
+
+    @internalProperty() private error = "";
+
+    @internalProperty() private COLS: Col[] = [];
+    @internalProperty() private ROWS: Cell[][] = [];
+    private updCols = () => this.requestUpdate("COLS");
+    private updRows = () => this.requestUpdate("ROWS");
+
+    @internalProperty() private dragRow = -1;
+    @internalProperty() private dropRow = -1;
+    @internalProperty() private drops: [number, number][] = [];
+    private dragRowElem: HTMLElement | null = null;
+
+    @internalProperty() private dragCol = -1;
+    @internalProperty() private dropCol = -1;
+
+    private isResizing = false;
+    private isSelectable = false;
+    private selected: Record<number, boolean> = {};
+    private expandedRowIdx: Record<number, boolean> = {};
+
+    private dragover = (e: Event) => e.preventDefault();
+
+    disconnectedCallback() {
+      super.disconnectedCallback();
+      document.removeEventListener("dragover", this.dragover);
+    }
+
+    connectedCallback() {
+      super.connectedCallback();
+      document.addEventListener("dragover", this.dragover);
+
+      // cols
+
+      let index = 0;
+      const pushCol = (col: ColOptions, group?: { name: string; length: number }) => {
+        const f = col.filters || this.config.default?.col?.filters;
+        const filters = f
+          ? f == "forString"
+            ? Filter.optionsString
+            : f == "forNumber"
+              ? Filter.optionsNumber
+              : f.length
+                ? f
+                : null
+          : null;
+
+        this.COLS.push({
+          options: { ...col },
+          index: index++,
+          group,
+          sort: "default",
+          sorter: col.sorter || this.config.default?.col?.sorter,
+          isCollapsable: this.config.cols.collapse == col.id,
+          filter: filters
+            ? {
+              list: filters,
+              selectedIndex: 0,
+              input: "",
+              active: false,
+              menuVisible: false
+            }
+            : undefined
+        });
+      };
+
+      this.config.cols.define.forEach(col => {
```

Comment:
why won't you just use an array reducer? especially with incrementing index the way you do above

Reply:
@yananym it is the way I think it must be done
with reducer - code will not become shorter or cleaner IMHO